exec/internal/compile/allocator.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2019 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build !appengine
+
+package compile
+
+import (
+	"unsafe"
+
+	mmap "github.com/edsrzf/mmap-go"
+)
+
+const (
+	minAllocSize = 1024
+	// alignment - instruction caching works better on aligned boundaries.
+	allocationAlignment = 128 - 1
+)
+
+type mmapBlock struct {
+	mem       mmap.MMap
+	consumed  uint32
+	remaining uint32
+}
+
+// MMapAllocator copies instructions into executable memory.
+type MMapAllocator struct {
+	last   *mmapBlock
+	blocks []*mmapBlock
+}
+
+// Close frees all pages allocted by the allocator.
+func (a *MMapAllocator) Close() error {
+	for _, block := range a.blocks {
+		if err := block.mem.Unmap(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AllocateExec allocates a block of executable memory with the given code contained.
+func (a *MMapAllocator) AllocateExec(asm []byte) (NativeCodeUnit, error) {
```

Comment:
do you foresee a strategy to reclaim mmap allocated blocks?
(just asking. I understand this can be quite tricky and might require a fair amount of additional infrastructure...)

Reply:
We are keeping a slice of the allocated blocks. The plan was to iterate through the first x blocks to see if there is any free space we can use instead of allocating a new page. I saved this work for a future PR.

All pages are freed on a call to `MMapAllocator.Close()`, which is called on a call to `vm.Close()`.